huaweicloud/services/dataarts/resource_huaweicloud_dataarts_architecture_data_standard_template.go

```diff
@@ -0,0 +1,632 @@
+	if err != nil {
+		return diag.Errorf("error creating DataArts Studio client: %s", err)
+	}
+
+	getDataStandardTemplateRespBody, err := getDataStandardTemplate(d, createDataStandardTemplateClient)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	hasTemplate := utils.PathSearch("data.value.hasTemplate", getDataStandardTemplateRespBody, false).(bool)
+	if hasTemplate {
+		return diag.Errorf("a DataArts Architecture data standard template has exist")
+	}
+
+	systemDefaultFields := utils.PathSearch("data.value.preFields_system_default", getDataStandardTemplateRespBody,
+		make([]interface{}, 0)).([]interface{})
+	optionalFields := utils.PathSearch("data.value.preFields_optional", getDataStandardTemplateRespBody,
+		make([]interface{}, 0)).([]interface{})
+
+	// init the template
+	initDataStandardTemplateRespBody, err := initDataStandardTemplate(d, createDataStandardTemplateClient,
+		buildInitDataStandardTemplateBodyParams(d, systemDefaultFields, optionalFields))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	d.SetId(d.Get("workspace_id").(string))
+
+	// for the optional fields, the value of required, searchable can not be set for the first initialization operation,
+	// it will be modified by subsequent initialization operations, so it is necessary to init again to set required,
+	// searchable of the optional fields
+	rawOptionalFields := d.Get("optional_fields").(*schema.Set).List()
+	if len(rawOptionalFields) > 0 {
+		fields := utils.PathSearch("data.value", initDataStandardTemplateRespBody, make([]interface{}, 0)).([]interface{})
+		templateNameToIdMap := buildTemplateNameToIdMap(fields)
+		updateOptionalFields := buildUpdateDataStandardTemplateRequestBodyOptionalField(
+			d.Get("optional_fields").(*schema.Set).List(), nil, templateNameToIdMap)
+		_, err = initDataStandardTemplate(d, createDataStandardTemplateClient,
+			map[string]interface{}{"fields": updateOptionalFields})
+		if err != nil {
+			return diag.FromErr(err)
+		}
+	}
+
+	return resourceDataStandardTemplateRead(ctx, d, meta)
+}
+
+func buildInitDataStandardTemplateBodyParams(d *schema.ResourceData, systemDefaultFields,
+	optionalFields []interface{}) map[string]interface{} {
+	fields := make([]map[string]interface{}, 0)
+	fields = append(fields, buildSystemDefaultOrOptionalFields(systemDefaultFields, true)...)
+	fields = append(fields, buildSystemDefaultOrOptionalFields(optionalFields, false)...)
+	fields = append(fields, buildCreateDataStandardTemplateRequestBodyCustomField(
+		d.Get("custom_fields").(*schema.Set).List())...)
+	return map[string]interface{}{"fields": fields}
+}
+
+func buildSystemDefaultOrOptionalFields(fields []interface{}, isSystemField bool) []map[string]interface{} {
+	if len(fields) == 0 {
+		return nil
+	}
+
+	rst := make([]map[string]interface{}, len(fields))
+	for i, v := range fields {
+		if raw, ok := v.(map[string]interface{}); ok {
+			rst[i] = map[string]interface{}{
```

Comment:
Please use `append` to replace `index`. This line of code may cause unexpected problems when `ok` is false.

Reply:
done